sql/expression/function/aggregation/unary_agg_buffers.go

```diff
@@ -666,3 +667,228 @@ func (j *jsonArrayBuffer) Eval(ctx *sql.Context) (interface{}, error) {
+	if s.count == 1 {
+		s.mean = val
+		return nil
+	}
+
+	newMean := calcOnlineMean(s.mean, val, s.count)
+	s.std2 = calcOnlineVar2(s.mean, newMean, s.std2, val)
+	s.mean = newMean
+
+	return nil
+}
+
+// Eval implements the AggregationBuffer interface.
+func (s *stdDevPopBuffer) Eval(ctx *sql.Context) (interface{}, error) {
+	if s.count == 0 {
+		return nil, nil
+	}
+	return math.Sqrt(s.std2 / float64(s.count)), nil
+}
+
+// Dispose implements the Disposable interface.
+func (s *stdDevPopBuffer) Dispose() {}
+
+type stdDevSampBuffer struct {
+	vals []interface{}
+	expr sql.Expression
+
+	count uint64
+	mean  float64
+	std2  float64
+}
+
+func NewStdDevSampBuffer(child sql.Expression) *stdDevSampBuffer {
+	return &stdDevSampBuffer{
+		vals: nil,
+		expr: child,
+	}
+}
+
+// Update implements the AggregationBuffer interface.
+func (s *stdDevSampBuffer) Update(ctx *sql.Context, row sql.Row) error {
+	v, err := evalFloat64(ctx, row, s.expr)
+	if err != nil {
+		return err
+	}
+	if v == nil {
+		return nil
+	}
+	val := v.(float64)
+
+	s.count += 1
+	if s.count == 1 {
+		s.mean = val
+		return nil
+	}
+
+	newMean := calcOnlineMean(s.mean, val, s.count)
+	s.std2 = calcOnlineVar2(s.mean, newMean, s.std2, val)
+	s.mean = newMean
+
+	return nil
+}
+
+// Eval implements the AggregationBuffer interface.
+func (s *stdDevSampBuffer) Eval(ctx *sql.Context) (interface{}, error) {
+	if s.count <= 1 {
+		return nil, nil
+	}
+	return math.Sqrt(s.std2 / float64(s.count-1)), nil
+}
+
+// Dispose implements the Disposable interface.
+func (s *stdDevSampBuffer) Dispose() {}
+
+type varPopBuffer struct {
+	vals []interface{}
+	expr sql.Expression
+
+	count uint64
+	mean  float64
+	std2  float64
+}
+
+func NewVarPopBuffer(child sql.Expression) *varPopBuffer {
+	return &varPopBuffer{
+		vals: nil,
+		expr: child,
+	}
+}
+
+// Update implements the AggregationBuffer interface.
+func (vp *varPopBuffer) Update(ctx *sql.Context, row sql.Row) error {
+	v, err := evalFloat64(ctx, row, vp.expr)
+	if err != nil {
+		return err
+	}
+	if v == nil {
+		return nil
+	}
+	val := v.(float64)
+
+	vp.count += 1
+	if vp.count == 1 {
+		vp.mean = val
+		return nil
+	}
+
+	newMean := calcOnlineMean(vp.mean, val, vp.count)
+	vp.std2 = calcOnlineVar2(vp.mean, newMean, vp.std2, val)
+	vp.mean = newMean
+
+	return nil
+}
+
+// Eval implements the AggregationBuffer interface.
+func (vp *varPopBuffer) Eval(ctx *sql.Context) (interface{}, error) {
+	if vp.count == 0 {
+		return nil, nil
+	}
+	return vp.std2 / float64(vp.count), nil
+}
+
+// Dispose implements the Disposable interface.
+func (vp *varPopBuffer) Dispose() {}
+
+type varSampBuffer struct {
+	vals []interface{}
+	expr sql.Expression
+
+	count uint64
+	mean  float64
+	std2  float64
+}
+
+func NewVarSampBuffer(child sql.Expression) *varSampBuffer {
+	return &varSampBuffer{
+		vals: nil,
+		expr: child,
+	}
+}
+
+// Update implements the AggregationBuffer interface.
+func (vp *varSampBuffer) Update(ctx *sql.Context, row sql.Row) error {
```

Comment:
are all of the `Update`s the same? could we have a single `varBuffer` implementation that the user-facing expressions inherit and overwrite with custom `Eval`s? same for window side. I imagine the testing surface area would shrink a lot

Reply:
makes sense, thanks!